File: apps/node/src/ask.ts
// "Ask your node": the FAQ-accordion overlay. Every answer renders a headline
// number, one plain-language line, and a citation pill — data comes from the
// ask_* Tauri commands (the local node's RPC or pure consensus math), never an
// external service. Commands resolve to a tagged Ask state; nothing throws.

import { invoke } from "@tauri-apps/api/core";

// ── Wire types (mirror apps/node/src-tauri/src/ask.rs) ──────────────────────

type Ask<T> =
  | { state: "ready"; data: T }
  | { state: "stopped" }
  | { state: "warming" }
  | { state: "unavailable"; data: { message: string } };

interface ChainProgress {
  height: number;
  headers: number;
  progress: number;
  near_tip: boolean;
  peers: number;
}
interface SupplyAnswer {
  mined_btx: number;
  cap_btx: number;
  pct: number;
  height: number;
}
interface HalvingAnswer {
  blocks_remaining: number;
  at_height: number;
  est_secs: number;
  from_reward_btx: number;
  to_reward_btx: number;
  height: number;
}
interface FeesAnswer {
  feerate_btx_kvb: number | null;
  mempool_txs: number;
  mempool_vsize: number;
}
interface MiningAnswer {
  difficulty: number;
  network_hashps: number;
  height: number;
}
interface BlockAnswer {
  height: number;
  hash: string;
  time: number;
  n_tx: number;
  size: number;
}
type TxLookup =
  | {
      kind: "found";
      txid: string;
      confirmations: number;
      block_height: number | null;
      block_time: number | null;
      vsize: number;
      vin_count: number;
      vout_count: number;
      total_out_btx: number;
    }
  | { kind: "needs_index" }
  | { kind: "building"; pct: number }
  | { kind: "not_found" };
interface TxIndexAnswer {
  enabled: boolean;
  configured: boolean;
  synced: boolean;
  pct: number;
}

// ── Free-text router (pure, unit-tested) ────────────────────────────────────

export type AskQueryKind = "empty" | "height" | "hash_or_txid" | "invalid";

export function classifyAskQuery(q: string): { kind: AskQueryKind } {
  const t = q.trim().replace(/,/g, "");
  if (t === "") return { kind: "empty" };
  if (/^\d+$/.test(t)) return { kind: "height" };
  if (/^[0-9a-fA-F]{64}$/.test(t)) return { kind: "hash_or_txid" };
  return { kind: "invalid" };
}

// ── Formatting ───────────────────────────────────────────────────────────────

const fmtInt = (n: number) => n.toLocaleString("en-US");
const fmtBtx = (n: number) =>
  n.toLocaleString("en-US", { maximumFractionDigits: n >= 1000 ? 0 : 4 });
const fmtWhen = (unixSecs: number) => new Date(unixSecs * 1000).toLocaleString();
/** ~90 s blocks → a rough, honest date ("around March 2027"). */
function fmtEta(estSecs: number): string {
  const d = new Date(Date.now() + estSecs * 1000);
  return d.toLocaleDateString("en-US", { year: "numeric", month: "long" });
}
const esc = (s: string) =>
  s.replace(/[&<>"']/g, (c) =>
    ({ "&": "&amp;", "<": "&lt;", ">": "&gt;", '"': "&quot;", "'": "&#39;" })[c] as string
  );

// ── Rendering ────────────────────────────────────────────────────────────────

const $ = <T extends HTMLElement = HTMLElement>(id: string): T =>
  document.getElementById(id) as T;

function cite(source: string): string {
  return `<div class="ask-cite"><span class="cite-dot"></span> via ${esc(source)}</div>`;
}

function answerHtml(headline: string, line: string, source: string): string {
  return (
    `<div class="ask-headline">${headline}</div>` +
    `<div class="ask-line">${line}</div>` +
    cite(source)
  );
}

function stateHtml(a: Ask<unknown>): string | null {
  switch (a.state) {
    case "stopped":
      return `<div class="ask-line">Start your node to ask it questions.</div>`;
    case "warming":
      return `<div class="ask-line">Your node is still catching up — ask again in a moment.</div>`;
    case "unavailable":
      return `<div class="ask-line">${esc(a.data.message)}</div>`;
    default:
      return null;
  }
}

// Renderers per question — each returns the .ask-a innerHTML.

function renderChain(a: Ask<ChainProgress>): string {
  const s = stateHtml(a);
  if (s) return s;
  const d = (a as { state: "ready"; data: ChainProgress }).data;
  const pct = Math.min(100, d.progress * 100);
  const head = d.near_tip
    ? `Block ${fmtInt(d.height)} — at the tip`
    : `Block ${fmtInt(d.height)} of ~${fmtInt(d.headers)}`;
  const line = d.near_tip
    ? `Your copy of the chain is current, connected to ${fmtInt(d.peers)} peer${d.peers === 1 ? "" : "s"}.`
    : `Catching up — about ${pct.toFixed(1)}% verified so far.`;
  return answerHtml(head, line, "getblockchaininfo + getchainstates");
}

function renderSupply(a: Ask<SupplyAnswer>): string {
  const s = stateHtml(a);
  if (s) return s;
  const d = (a as { state: "ready"; data: SupplyAnswer }).data;
  return answerHtml(
    `${fmtBtx(d.mined_btx)} BTX`,
    `Mined so far, of the ${fmtBtx(d.cap_btx)} cap — ${(d.pct * 100).toFixed(1)}%, from the subsidy schedule at block ${fmtInt(d.height)}.`,
    "block height × subsidy schedule"
  );
}

function renderHalving(a: Ask<HalvingAnswer>): string {
  const s = stateHtml(a);
  if (s) return s;
  const d = (a as { state: "ready"; data: HalvingAnswer }).data;
  return answerHtml(
    `${fmtInt(d.blocks_remaining)} blocks to go`,
    `At block ${fmtInt(d.at_height)} the reward goes ${fmtBtx(d.from_reward_btx)} → ${fmtBtx(d.to_reward_btx)} BTX — around ${fmtEta(d.est_secs)} at 90-second blocks.`,
    "block height math"
  );
}

function renderFees(a: Ask<FeesAnswer>): string {
  const s = stateHtml(a);
  if (s) return s;
  const d = (a as { state: "ready"; data: FeesAnswer }).data;
  if (d.feerate_btx_kvb === null) {
    return answerHtml(
      d.mempool_txs === 0 ? "Quiet right now" : `${fmtInt(d.mempool_txs)} waiting`,
      d.mempool_txs === 0
        ? "The mempool is empty — not enough recent activity for a fee estimate."
        : `Not enough recent activity for an estimate; the mempool holds ${fmtInt(d.mempool_txs)} transactions (${fmtInt(d.mempool_vsize)} vB).`,
      "getmempoolinfo"
    );
  }
  return answerHtml(
    `${(d.feerate_btx_kvb * 100_000).toFixed(2)} sats/vB`,
    `Estimated for ~6-block confirmation; ${fmtInt(d.mempool_txs)} transaction${d.mempool_txs === 1 ? "" : "s"} waiting.`,
    "estimatesmartfee + getmempoolinfo"
  );
}

function renderMining(a: Ask<MiningAnswer>): string {
  const s = stateHtml(a);
  if (s) return s;
  const d = (a as { state: "ready"; data: MiningAnswer }).data;
  const hs =
    d.network_hashps >= 1e6
      ? `${(d.network_hashps / 1e6).toFixed(1)}M`
      : fmtInt(Math.round(d.network_hashps));
  return answerHtml(
    `Difficulty ${d.difficulty.toLocaleString("en-US", { maximumFractionDigits: 2 })}`,
    `The network is solving about ${hs} matrix problems per second (an estimate from recent blocks, at height ${fmtInt(d.height)}).`,
    "getmininginfo"
  );
}

function renderBlock(a: Ask<BlockAnswer>): string {
  const s = stateHtml(a);
  if (s) return s;
  const d = (a as { state: "ready"; data: BlockAnswer }).data;
  return answerHtml(
    `Block ${fmtInt(d.height)}`,
    `${fmtWhen(d.time)} · ${fmtInt(d.n_tx)} transaction${d.n_tx === 1 ? "" : "s"} · ${fmtInt(d.size)} bytes<br><span class="ask-hash">${esc(d.hash)}</span>`,
    "getblockhash + getblock"
  );
}

function renderTx(a: Ask<TxLookup>): string {
  const s = stateHtml(a);
  if (s) return s;
  const d = (a as { state: "ready"; data: TxLookup }).data;
  switch (d.kind) {
    case "found": {
      const conf =
        d.confirmations === 0
          ? "In the mempool — not confirmed yet"
          : `${fmtInt(d.confirmations)} confirmation${d.confirmations === 1 ? "" : "s"}`;
      const where = d.block_height !== null ? ` · block ${fmtInt(d.block_height)}` : "";
      const when = d.block_time !== null ? ` · ${fmtWhen(d.block_time)}` : "";
      return answerHtml(
        conf,
        `${d.vin_count} in → ${d.vout_count} out · ${fmtBtx(d.total_out_btx)} BTX total moved · ${fmtInt(d.vsize)} vB${where}${when}`,
        "getrawtransaction"
      );
    }
    case "building":
      return `<div class="ask-line">Building the transaction index… ${(d.pct * 100).toFixed(0)}% — your node stays running; this works when it finishes.</div>`;
    case "needs_index":
      return (
        `<div class="ask-line">Historical lookups need <strong>Explorer mode</strong>: your node builds a one-time transaction index in the background (extra disk, node stays usable). Mempool transactions work without it.</div>` +
        `<button id="ask-enable-explorer" class="btn-secondary" type="button">Turn on Explorer mode</button>`
      );
    case "not_found":
      return `<div class="ask-line">No transaction with that id on your node.</div>`;
  }
}

// ── Wiring ───────────────────────────────────────────────────────────────────

type AskKey = "chain" | "supply" | "halving" | "fees" | "mining" | "block" | "tx";

async function fetchAnswer(key: AskKey, arg?: string): Promise<string> {
  try {
    switch (key) {
      case "chain":
        return renderChain(await invoke<Ask<ChainProgress>>("ask_chain_progress"));
      case "supply":
        return renderSupply(await invoke<Ask<SupplyAnswer>>("ask_supply"));
      case "halving":
        return renderHalving(await invoke<Ask<HalvingAnswer>>("ask_next_halving"));
      case "fees":
        return renderFees(await invoke<Ask<FeesAnswer>>("ask_fees"));
      case "mining":
        return renderMining(await invoke<Ask<MiningAnswer>>("ask_mining"));
      case "block":
        return renderBlock(await invoke<Ask<BlockAnswer>>("ask_block", { query: arg ?? "" }));
      case "tx": {
        if (!arg) {
          return `<div class="ask-line">Paste a transaction id (64 hex characters) in the box below, then Look up.</div>`;
        }
        return renderTx(await invoke<Ask<TxLookup>>("ask_transaction", { txid: arg }));
      }
    }
  } catch (e) {
    return `<div class="ask-line">${esc(String(e))}</div>`;
  }
}

export function initAsk(): void {
  const overlay = $("ask-overlay");
  let indexPoll: ReturnType<typeof setInterval> | undefined;

  const rowOf = (key: AskKey) =>
    overlay.querySelector<HTMLElement>(`.ask-item[data-ask="${key}"]`)!;

  function collapse(item: HTMLElement): void {
    item.querySelector(".ask-q")!.setAttribute("aria-expanded", "false");
    (item.querySelector(".ask-a") as HTMLElement).hidden = true;
  }

  async function expand(key: AskKey, arg?: string): Promise<void> {
    const item = rowOf(key);
    const btn = item.querySelector(".ask-q")!;
    const body = item.querySelector(".ask-a") as HTMLElement;
    btn.setAttribute("aria-expanded", "true");
    body.hidden = false;
    body.innerHTML = `<div class="ask-line ask-loading">Asking your node…</div>`;
    body.innerHTML = await fetchAnswer(key, arg);
    wireExplorerButton(body);
  }

  // The just-in-time Explorer-mode prompt lives inside the tx answer.
  function wireExplorerButton(scope: HTMLElement): void {
    const btn = scope.querySelector<HTMLButtonElement>("#ask-enable-explorer");
    if (!btn) return;
    btn.addEventListener("click", async () => {
      btn.disabled = true;
      btn.textContent = "Turning on… (restarts the node)";
      try {
        await invoke("set_explorer_mode", { on: true });
        startIndexPoll();
      } catch (e) {
        btn.disabled = false;
        btn.textContent = "Turn on Explorer mode";
        scope.insertAdjacentHTML(
          "beforeend",
          `<div class="ask-line">${esc(String(e))}</div>`
        );
      }
    });
  }

  // While the index builds, keep the tx row's progress line fresh.
  function startIndexPoll(): void {
    stopIndexPoll();
    const body = rowOf("tx").querySelector(".ask-a") as HTMLElement;
    const tickOnce = async () => {
      try {
        const st = await invoke<Ask<TxIndexAnswer>>("ask_tx_index_status");
        if (st.state !== "ready") return;
        if (st.data.synced) {
          stopIndexPoll();
          body.innerHTML = `<div class="ask-line">Explorer mode is ready — paste a transaction id below and Look up.</div>`;
          return;
        }
        body.innerHTML = `<div class="ask-line">Building the transaction index… ${(st.data.pct * 100).toFixed(0)}% — your node stays running.</div>`;
      } catch {
        /* keep polling; transient during the restart */
      }
    };
    void tickOnce();
    indexPoll = setInterval(() => void tickOnce(), 3000);
  }
  function stopIndexPoll(): void {
    if (indexPoll !== undefined) clearInterval(indexPoll);
    indexPoll = undefined;
  }

  // Accordion: tap toggles; expanding re-fetches (fresh numbers every open).
  overlay.querySelectorAll<HTMLElement>(".ask-item").forEach((item) => {
    const key = item.dataset.ask as AskKey;
    item.querySelector(".ask-q")!.addEventListener("click", () => {
      const expanded =
        item.querySelector(".ask-q")!.getAttribute("aria-expanded") === "true";
      if (expanded) collapse(item);
      else void expand(key);
    });
  });

  // Free-text router: height → block row; 64-hex → block first, then tx.
  async function lookup(): Promise<void> {
    const input = $<HTMLInputElement>("ask-search");
    const q = input.value;
    const cls = classifyAskQuery(q);
    if (cls.kind === "empty") return;
    if (cls.kind === "invalid" || cls.kind === "height") {
      // ask_block answers invalid input with its calm one-line explanation.
      await expand("block", q);
      return;
    }
    // 64-hex: try a block hash first; fall through to a tx lookup.
    const blockAns = await invoke<Ask<BlockAnswer>>("ask_block", { query: q.trim() });
    if (blockAns.state === "ready") {
      const item = rowOf("block");
      item.querySelector(".ask-q")!.setAttribute("aria-expanded", "true");
      const body = item.querySelector(".ask-a") as HTMLElement;
      body.hidden = false;
      body.innerHTML = renderBlock(blockAns);
      return;
    }
    await expand("tx", q.trim().toLowerCase());
  }
  $("ask-search-btn").addEventListener("click", () => void lookup());
  $<HTMLInputElement>("ask-search").addEventListener("keydown", (e) => {
    if (e.key === "Enter") void lookup();
  });

  // Node-stopped gate: probed once per open with the lightest command
  // (ask_supply → one getchainstates call; pure math otherwise).
  async function refreshGate(): Promise<void> {
    try {
      const probe = await invoke<Ask<SupplyAnswer>>("ask_supply");
      $("ask-gate").hidden = probe.state !== "stopped";
    } catch {
      $("ask-gate").hidden = true; // outside Tauri (browser QA) — no gate
    }
  }
  $("ask-start-btn").addEventListener("click", async () => {
    const b = $<HTMLButtonElement>("ask-start-btn");
    b.disabled = true;
    b.textContent = "Starting…";
    try {
      await invoke("start_node");
      await refreshGate();
    } catch (e) {
      console.error("start from ask panel failed", e);
    } finally {
      b.disabled = false;
      b.textContent = "Start node";
    }
  });

  // Open / close.
  $("ask-btn").addEventListener("click", () => {
    overlay.hidden = false;
    void refreshGate();
  });
  $("ask-close").addEventListener("click", () => {
    overlay.hidden = true;
    stopIndexPoll();
  });
  overlay.addEventListener("click", (e) => {
    if (e.target === overlay) {
      overlay.hidden = true;
      stopIndexPoll();
    }
  });

  $("ask-global-stats").addEventListener("click", () => {
    void invoke("open_global_stats").catch(() => {});
  });

  // QA hook (harmless in production, mirrors __ambient/__core).
  (window as unknown as Record<string, unknown>).__ask = {
    open: () => {
      overlay.hidden = false;
    },
  };
}

File: apps/node/src-tauri/src/ask.rs
//! "Ask your node" — calm, cited answers straight from the user's own node.
//!
//! Every command resolves (never rejects) to an [`Ask`] state: the panel
//! renders `stopped` / `warming` / `unavailable` as plain sentences, so no
//! error ever throws to the webview. Numbers come from the local node's RPC
//! or pure consensus math (`btx_core::supply`) — no external service.

use serde::Serialize;
use tauri::{AppHandle, State};

use btx_core::error::AppError;
use btx_core::node_api as api;
use btx_core::rpc::RpcClient;
use btx_core::supply;

use crate::commands::{snapshot_spec, start_node_inner, stop_node_inner};
use crate::state::{node_datadir, AppState, NodeAppSettings};

/// Wire shape: `{state: "ready", data: T} | {state: "stopped"} | …`.
#[derive(Debug, Clone, Serialize)]
#[serde(tag = "state", content = "data", rename_all = "snake_case")]
pub enum Ask<T: Serialize> {
    Ready(T),
    /// Node not running — the panel shows "Start your node to ask it questions."
    Stopped,
    /// RPC_IN_WARMUP (-28): alive but catching up — "ask again in a moment".
    Warming,
    /// Anything else, as one plain sentence.
    Unavailable {
        message: String,
    },
}

/// Map an RPC failure to its calm state.
pub(crate) fn degrade<T: Serialize>(e: AppError) -> Ask<T> {
    match e {
        AppError::Rpc { code: -28, .. } => Ask::Warming,
        other => Ask::Unavailable {
            message: other.to_string(),
        },
    }
}

/// The armed RPC client, or `None` when the node is stopped.
async fn rpc_handle(state: &State<'_, AppState>) -> Option<RpcClient> {
    state.rpc.lock().await.clone()
}

/// The height the user experiences: snapshot-chainstate-aware (assumeutxo).
async fn best_height(rpc: &RpcClient) -> Result<u64, AppError> {
    let chainstates = api::get_chainstates(rpc).await?;
    let h = chainstates.best_height();
    if h > 0 {
        return Ok(h);
    }
    // Degenerate/older shape: fall back to getblockchaininfo.
    Ok(api::get_blockchain_info(rpc).await?.blocks)
}

// ── 1. How far along is the chain? ──────────────────────────────────────────

#[derive(Debug, Clone, Serialize)]
pub struct ChainProgress {
    pub height: u64,
    pub headers: u64,
    pub progress: f64,
    pub near_tip: bool,
    pub peers: i64,
}

#[tauri::command]
pub async fn ask_chain_progress(state: State<'_, AppState>) -> Result<Ask<ChainProgress>, String> {
    let Some(rpc) = rpc_handle(&state).await else {
        return Ok(Ask::Stopped);
    };
    let chain = match api::get_blockchain_info(&rpc).await {
        Ok(c) => c,
        Err(e) => return Ok(degrade(e)),
    };
    let (chainstates, peers) =
        tokio::join!(api::get_chainstates(&rpc), api::get_connection_count(&rpc));
    let chainstates = chainstates.unwrap_or_default();
    let readiness =
        btx_core::health::sync_readiness(&chainstates, &chain, snapshot_spec().anchor_height);
    Ok(Ask::Ready(ChainProgress {
        height: readiness.height(),
        headers: chain.headers,
        progress: readiness.progress(),
        near_tip: readiness.is_near_tip(),
        peers: peers.unwrap_or(0),
    }))
}

// ── 2. How much BTX exists so far? (pure math from height) ──────────────────

#[derive(Debug, Clone, Serialize)]
pub struct SupplyAnswer {
    pub mined_btx: f64,
    pub cap_btx: f64,
    pub pct: f64,
    pub height: u64,
}

#[tauri::command]
pub async fn ask_supply(state: State<'_, AppState>) -> Result<Ask<SupplyAnswer>, String> {
    let Some(rpc) = rpc_handle(&state).await else {
        return Ok(Ask::Stopped);
    };
    match best_height(&rpc).await {
        Ok(height) => {
            let mined_btx = supply::sats_to_btx(supply::mined_supply_sats(height));
            Ok(Ask::Ready(SupplyAnswer {
                mined_btx,
                cap_btx: supply::SUPPLY_CAP_BTX,
                pct: mined_btx / supply::SUPPLY_CAP_BTX,
                height,
            }))
        }
        Err(e) => Ok(degrade(e)),
    }
}

// ── 3. When's the next halving? ──────────────────────────────────────────────

#[derive(Debug, Clone, Serialize)]
pub struct HalvingAnswer {
    pub blocks_remaining: u64,
    pub at_height: u64,
    pub est_secs: u64,
    pub from_reward_btx: f64,
    pub to_reward_btx: f64,
    pub height: u64,
}

#[tauri::command]
pub async fn ask_next_halving(state: State<'_, AppState>) -> Result<Ask<HalvingAnswer>, String> {
    let Some(rpc) = rpc_handle(&state).await else {
        return Ok(Ask::Stopped);
    };
    match best_height(&rpc).await {
        Ok(height) => {
            let h = supply::next_halving(height);
            Ok(Ask::Ready(HalvingAnswer {
                blocks_remaining: h.blocks_remaining,
                at_height: h.at_height,
                est_secs: h.est_secs,
                from_reward_btx: supply::sats_to_btx(h.from_subsidy_sats),
                to_reward_btx: supply::sats_to_btx(h.to_subsidy_sats),
                height,
            }))
        }
        Err(e) => Ok(degrade(e)),
    }
}

// ── 4. What are fees like right now? ─────────────────────────────────────────

#[derive(Debug, Clone, Serialize)]
pub struct FeesAnswer {
    /// BTX per kvB for ~6-block confirmation; `null` = the node has no
    /// estimate (quiet network) — the panel falls back to the mempool line.
    pub feerate_btx_kvb: Option<f64>,
    pub mempool_txs: u64,
    pub mempool_vsize: u64,
}

#[tauri::command]
pub async fn ask_fees(state: State<'_, AppState>) -> Result<Ask<FeesAnswer>, String> {
    let Some(rpc) = rpc_handle(&state).await else {
        return Ok(Ask::Stopped);
    };
    let mempool = match api::get_mempool_info(&rpc).await {
        Ok(m) => m,
        Err(e) => return Ok(degrade(e)),
    };
    // No estimate is a normal quiet-network answer, never an error.
    let feerate = api::estimate_smart_fee(&rpc, 6).await.unwrap_or(None);
    Ok(Ask::Ready(FeesAnswer {
        feerate_btx_kvb: feerate,
        mempool_txs: mempool.size,
        mempool_vsize: mempool.bytes,
    }))
}

// ── 5. How hard is mining right now? ─────────────────────────────────────────

#[derive(Debug, Clone, Serialize)]
pub struct MiningAnswer {
    pub difficulty: f64,
    pub network_hashps: f64,
    pub height: u64,
}

#[tauri::command]
pub async fn ask_mining(state: State<'_, AppState>) -> Result<Ask<MiningAnswer>, String> {
    let Some(rpc) = rpc_handle(&state).await else {
        return Ok(Ask::Stopped);
    };
    match api::get_mining_info(&rpc).await {
        Ok(mi) => Ok(Ask::Ready(MiningAnswer {
            difficulty: mi.difficulty,
            network_hashps: mi.network_hashps,
            height: mi.blocks,
        })),
        Err(e) => Ok(degrade(e)),
    }
}

// ── 6. Show me a block ───────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize)]
pub struct BlockAnswer {
    pub height: u64,
    pub hash: String,
    pub time: u64,
    pub n_tx: u64,
    pub size: u64,
}

/// `query`: empty → the tip block; digits → that height; 64-hex → that hash.
#[tauri::command]
pub async fn ask_block(
    state: State<'_, AppState>,
    query: String,
) -> Result<Ask<BlockAnswer>, String> {
    let Some(rpc) = rpc_handle(&state).await else {
        return Ok(Ask::Stopped);
    };
    let q = query.trim().replace(',', "");
    let fetched = if q.is_empty() {
        match best_height(&rpc).await {
            Ok(h) => api::get_block_by_height(&rpc, h).await,
            Err(e) => Err(e),
        }
    } else if q.chars().all(|c| c.is_ascii_digit()) {
        match q.parse::<u64>() {
            Ok(h) => api::get_block_by_height(&rpc, h).await,
            Err(_) => {
                return Ok(Ask::Unavailable {
                    message: "That block height is out of range.".into(),
                })
            }
        }
    } else if q.len() == 64 && q.chars().all(|c| c.is_ascii_hexdigit()) {
        api::get_block_by_hash(&rpc, &q).await
    } else {
        return Ok(Ask::Unavailable {
            message: "Enter a block height (a number) or a 64-character block hash.".into(),
        });
    };
    match fetched {
        Ok(b) => Ok(Ask::Ready(BlockAnswer {
            height: b.height,
            hash: b.hash,
            time: b.time,
            n_tx: b.n_tx,
            size: b.size,
        })),
        Err(e) => Ok(block_error_to_ask(e)),
    }
}

/// Map a block-lookup failure to its calm sentence. Split out for testing:
/// besides -5/-8 (unknown height/hash), a FRESHLY fast-started node answers
/// -1 "Block not available (not fully downloaded)" for most historical
/// heights — assumeutxo loads the UTXO set instantly but the raw blocks only
/// arrive as the background backfill progresses (observed live in the v0.2
/// e2e). That's a normal state, not an error.
fn block_error_to_ask(e: AppError) -> Ask<BlockAnswer> {
    match e {
        AppError::Rpc { code: -5, .. } | AppError::Rpc { code: -8, .. } => Ask::Unavailable {
            message: "No block with that height or hash on your node.".into(),
        },
        AppError::Rpc {
            code: -1,
            ref message,
        } if message.contains("Block not available") => Ask::Unavailable {
            message: "Your node hasn't downloaded that block yet — it's still \
                          backfilling older history in the background. Ask for a newer \
                          block, or try again later."
                .into(),
        },
        other => degrade(other),
    }
}

// ── Gated: look up a transaction (Explorer mode) ─────────────────────────────

#[derive(Debug, Clone, Serialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum TxLookup {
    Found {
        txid: String,
        confirmations: u64,
        block_height: Option<u64>,
        block_time: Option<u64>,
        vsize: u64,
        vin_count: usize,
        vout_count: usize,
        total_out_btx: f64,
    },
    /// Historical lookup needs Explorer mode (txindex) — the just-in-time
    /// prompt renders from this.
    NeedsIndex,
    /// txindex is on but still building — pct of the chain indexed so far.
    Building {
        pct: f64,
    },
    NotFound,
}

#[tauri::command]
pub async fn ask_transaction(
    state: State<'_, AppState>,
    txid: String,
) -> Result<Ask<TxLookup>, String> {
    let Some(rpc) = rpc_handle(&state).await else {
        return Ok(Ask::Stopped);
    };
    let t = txid.trim().to_lowercase();
    if t.len() != 64 || !t.chars().all(|c| c.is_ascii_hexdigit()) {
        return Ok(Ask::Unavailable {
            message: "A transaction id is 64 hex characters.".into(),
        });
    }
    match api::get_raw_transaction(&rpc, &t).await {
        Ok(v) => {
            let s = api::tx_summary(&v);
            // Enrich with the block height when the tx is confirmed (verbose
            // getrawtransaction carries the hash but not the height).
            let block_height = match &s.block_hash {
                Some(h) => api::get_block_by_hash(&rpc, h).await.ok().map(|b| b.height),
                None => None,
            };
            Ok(Ask::Ready(TxLookup::Found {
                txid: s.txid,
                confirmations: s.confirmations,
                block_height,
                block_time: s.block_time,
                vsize: s.vsize,
                vin_count: s.vin_count,
                vout_count: s.vout_count,
                total_out_btx: s.total_out_btx,
            }))
        }
        Err(AppError::Rpc { code: -5, .. }) => {
            // Not in the mempool. Without txindex a historical tx is
            // unreachable → the gate. With txindex: building vs truly absent.
            if !NodeAppSettings::load(&node_datadir()).txindex_enabled {
                return Ok(Ask::Ready(TxLookup::NeedsIndex));
            }
            match api::get_tx_index_info(&rpc).await {
                Ok(Some(st)) if !st.synced => {
                    let tip = best_height(&rpc).await.unwrap_or(0).max(1);
                    Ok(Ask::Ready(TxLookup::Building {
                        pct: st.best_block_height as f64 / tip as f64,
                    }))
                }
                _ => Ok(Ask::Ready(TxLookup::NotFound)),
            }
        }
        Err(e) => Ok(degrade(e)),
    }
}

// ── Explorer mode (txindex) ──────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize)]
pub struct TxIndexAnswer {
    /// The persisted user choice (settings).
    pub enabled: bool,
    /// Whether the RUNNING node has the index configured (getindexinfo).
    pub configured: bool,
    pub synced: bool,
    pub pct: f64,
}

#[tauri::command]
pub async fn ask_tx_index_status(state: State<'_, AppState>) -> Result<Ask<TxIndexAnswer>, String> {
    let enabled = NodeAppSettings::load(&node_datadir()).txindex_enabled;
    let Some(rpc) = rpc_handle(&state).await else {
        return Ok(Ask::Stopped);
    };
    match api::get_tx_index_info(&rpc).await {
        Ok(Some(st)) => {
            let tip = best_height(&rpc).await.unwrap_or(0).max(1);
            Ok(Ask::Ready(TxIndexAnswer {
                enabled,
                configured: true,
                synced: st.synced,
                pct: if st.synced {
                    1.0
                } else {
                    st.best_block_height as f64 / tip as f64
                },
            }))
        }
        Ok(None) => Ok(Ask::Ready(TxIndexAnswer {
            enabled,
            configured: false,
            synced: false,
            pct: 0.0,
        })),
        Err(e) => Ok(degrade(e)),
    }
}

/// Turn Explorer mode on/off: persist the choice, write/remove `txindex=1` in
/// the faststart conf, and (when the node is running) restart it gracefully so
/// btxd picks the flag up. The node stays fully usable while the index builds.
#[tauri::command]
pub async fn set_explorer_mode(
    app: AppHandle,
    state: State<'_, AppState>,
    on: bool,
) -> Result<(), String> {
    let datadir = node_datadir();
    let conf = datadir.join("faststart").join("faststart.conf");
    btx_core::setup::set_conf_kv(&conf, "txindex", if on { Some("1") } else { None })
        .map_err(|e| e.to_string())?;
    NodeAppSettings::update(&datadir, |s| s.txindex_enabled = on);
    if state.rpc.lock().await.is_some() {
        stop_node_inner(&state).await;
        start_node_inner(&app, &state).await?;
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn ask_serializes_tagged_states() {
        let ready = Ask::Ready(SupplyAnswer {
            mined_btx: 3_114_020.0,
            cap_btx: 21_000_000.0,
            pct: 0.1483,
            height: 155_700,
        });
        let v = serde_json::to_value(&ready).unwrap();
        assert_eq!(v["state"], "ready");
        assert_eq!(v["data"]["height"], 155_700);

        let stopped: Ask<SupplyAnswer> = Ask::Stopped;
        assert_eq!(serde_json::to_value(&stopped).unwrap()["state"], "stopped");

        let warm: Ask<SupplyAnswer> = degrade(AppError::Rpc {
            code: -28,
            message: "warming".into(),
        });
        assert_eq!(serde_json::to_value(&warm).unwrap()["state"], "warming");

        let un: Ask<SupplyAnswer> = degrade(AppError::Http("boom".into()));
        let v = serde_json::to_value(&un).unwrap();
        assert_eq!(v["state"], "unavailable");
        assert!(v["data"]["message"].as_str().unwrap().contains("boom"));
    }

    #[test]
    fn block_errors_map_to_calm_sentences() {
        // Unknown height/hash.
        let a = block_error_to_ask(AppError::Rpc {
            code: -5,
            message: "Block not found".into(),
        });
        match a {
            Ask::Unavailable { message } => assert!(message.contains("No block")),
            other => panic!("expected Unavailable, got {other:?}"),
        }
        // Fresh fast-started node: block data not backfilled yet.
        let a = block_error_to_ask(AppError::Rpc {
            code: -1,
            message: "Block not available (not fully downloaded)".into(),
        });
        match a {
            Ask::Unavailable { message } => {
                assert!(message.contains("backfilling"), "got: {message}")
            }
            other => panic!("expected Unavailable, got {other:?}"),
        }
        // Warmup still degrades to Warming, not a block message.
        let a = block_error_to_ask(AppError::Rpc {
            code: -28,
            message: "Verifying blocks".into(),
        });
        assert!(matches!(a, Ask::Warming));
    }

    #[test]
    fn tx_lookup_serializes_kinds() {
        let v = serde_json::to_value(&TxLookup::NeedsIndex).unwrap();
        assert_eq!(v["kind"], "needs_index");
        let v = serde_json::to_value(&TxLookup::Building { pct: 0.42 }).unwrap();
        assert_eq!(v["kind"], "building");
        assert_eq!(v["pct"], 0.42);
    }
}
